DietPi-Login | Failsafe hardening and resolving possible first run loops

MichaIng opened the pull request:
**Status**: Testing

**Reference**: https://github.com/Fourdee/DietPi/issues/2478

**Commit list/description**:
+ DietPi-Login | Remove "Filesystem prep" install state info, as this does not exist anymore, instead warn user about unknown install state, ask for check and report issue
+ DietPi-Login | Prompt G_WHIP_MSG, if install state is "-1" to make this more prominent
+ DietPi-Login | Check for root user before starting first run setup, otherwise ask user to login as root
+ DietPi-Login | Assure that user accepts license interactively one time: Skip on automated install to have it shown on first normal login instead
+ DietPi-Login | If on first run setup dietpi-update or dietpi-software is running in other session, show same G_WHIP once and exit script to prevent user from being kept in loop that can never be resolved in this session.
+ DietPi-Login | Re-read install state after first run update and dietpi-software run, to exit loop if immediate reboot was skipped. Allow update and installs in a single loop, e.g. if already on latest version, so second loop is an error indicator. Inform user in case. Apply install state 0, if not valid before second loop to prevent unlimited repeats.
+ DietPi-Boot | Assure it has finished before OpenSSH and Dropbear to prevent too early login
+ DietPi-Boot | Add "Before=getty-pre.target" which is reached on Buster before any getty starts. This is actually what we want, does not exist on Stretch yet but does not harm to add. On Buster we should add it to Wants= as well, as it is not pulled in automatically. But on Stretch this would produce an error, as it does not exist: https://www.freedesktop.org/software/systemd/man/systemd.special.html#getty-pre.target

MichaIng committed:
v6.22

+ DietPi-Login | Remove "Filesystem prep" install state info, as this does not exist anymore, instead warn user about unknown install state, ask for check and report issue
+ DietPi-Login | Prompt G_WHIP_MSG, if install state is "-1" to make this more prominent
+ DietPi-Login | Check for root user before starting first run setup, otherwise ask user to login as root
+ DietPi-Login | Assure that user accepts license interactively one time: Skip on automated install to have it shown on first normal login instead
+ DietPi-Login | If on first run setup dietpi-update or dietpi-software is running in other session, show same G_WHIP once and exit script to prevent user from being kept in loop that can never be resolved in this session.
+ DietPi-Login | Re-read install state after first run update and dietpi-software run, to exit loop if immediate reboot was skipped. Allow update and installs in a single loop, e.g. if already on latest version, so second loop is an error indicator. Inform user in case. Apply install state 0, if not valid before second loop to prevent unlimited repeats.

MichaIng commented on dietpi/login:
```diff
@@ -139,89 +134,92 @@
 
-						G_DIETPI-NOTIFY 2 'DietPi-Software is already running on another terminal/screen. Please complete or exit, before continuing.'
+				# - Start DietPi-Software
+				/DietPi/dietpi/dietpi-software | tee $FP_TMP_DIETPI_FIRSTRUNSETUP_LOG # Sets G_DIETPI_INSTALL_STAGE=2
```
@Fourdee 
If user chooses to reboot after `dietpi-software` has finished (most do, I guess), first run log will not be saved to disk currently.
Not sure how to solve best, but writing log to disk directly at least is an easy solution.

Fourdee replied:
@MichaIng 

Yep, need to write to disk for this 👍

MichaIng commented:
Actually I am not sure if the loop on first run setup is a good idea at all. A second attempt could make things worse or override initial errors with follow up errors that are harder to debug.
In general the loop should never occur anyway, since reboot after update and dietpi-software runs both is expected. Now if somehow the reboot after update is skipped (actually it is forced currently) software installs will be done in the same loop.

Also waiting for RAMdisk is obsolete. On local console, login is not possible before dietpi-boot has finished (Before=gettys) and via SSH not 100% sure, would need to check Before/After rules on that, especially on Jessie. But perhaps it would be good to add `Before=ssh.service dropbear.service` to DietPi-Boot, to prevent too early SSH logins which includes RAMdisk and PreBoot then as well?

PostBoot is okay as is, doesn't harm if services are not yet fully up on SSH login.

During Login at best we then check for successful exited DietPi-Boot (instead of RAMdisk), otherwise prompt info/error to user.

@Fourdee 
What you think?

Fourdee committed:
v6.22

+ Minor code
+ Log to disk

Commit:
v6.22

+ Minor optimization

MichaIng commented on dietpi/login:
```diff
@@ -196,16 +195,15 @@
 
 			# Check for second loop and in case inform user, since this indicates an error
-			if ! (( $G_DIETPI_INSTALL_STAGE == 2 )); then
+			if (( $G_DIETPI_INSTALL_STAGE != 2 )); then
```
@Fourdee 
Ah this was wanted. `! (( $G_DIETPI_INSTALL_STAGE == 2 ))` includes a check for valid integer.
- If somehow install state got corrupted empty or what, `! (( $G_DIETPI_INSTALL_STAGE == 2 ))` result is true, so error prompts and such is done.
- `(( $G_DIETPI_INSTALL_STAGE != 2 ))` in case of non-integer would result in false, thus the second loop would start silently.

Fourdee replied:
@MichaIng 

Ah apologies, will revert

Fourdee commented:
@MichaIng 

> What you think?

Yep, needs some changes, will take a look.

I've been tinkering with removing the reboots during 1st run setup, as we now have a system in update which reboots only if we required.

We then only need to reload globals after the update, only reboot then after dietpi-software install finished.

Commit:
v6.22

+ revert

MichaIng committed:
v6.22

+ DietPi-Boot | Assure it has finished before OpenSSH and Dropbear to prevent too early login
+ DietPi-Boot | Add "Before=getty-pre.target" which is reached on Buster before any getty starts. This is actually what we want, does not exist on Stretch yet but does not harm to add. On Buster we should add it to Wants= as well, as it is not pulled in automatically. But on Stretch this would produce an error, as it does not exist.

MichaIng committed:
v6.22

+ DietPi-Login | Log STDERR to first run setup log as well
+ DietPi-Login | Add info comment why we use "! (( $G_DIETPI_INSTALL_STAGE == 2 ))", to not forget

Commit:
v6.22

+ fix: Does not run on RPi during boot.

Commit:
Revert "v6.22"

This reverts commit 1467813234e7eb0e41cd6038c89f23d7e8bac0bd.

Commit:
v6.22

+ Fix: does not run on RPi during boot.

Commit:
v6.22

+ Login script rewrite
+ Minimize system reboots required aftert updates

Not finished yet, will test and patch.

Commit:
Merge branch 'login' of https://github.com/Fourdee/DietPi into login

Fourdee commented:
Will merge into dev and run some tests.

Fourdee commented:
@MichaIng 

Apologies, kind of took over your PR 🤣 

Been meaning to make these changes for a while, felt like a good time 😃

Fourdee commented:
@MichaIng 

In short, first run changes:
- Updates no longer reboot the system, unless we set them in ```patch_file``` or ```pre-patch_file```
- No reboot required, if user makes no changes to software selection in ```dietpi-software```

MichaIng commented on rootfs/etc/systemd/system/dietpi-postboot.service:
```diff
@@ -2,7 +2,7 @@
 # Order 4
 Requires=dietpi-boot.service
-After=dietpi-boot.service multi-user.target
+After=dietpi-boot.service
```
@Fourdee 
This was for the following reason:
- If `dietpi-postboot.service` runs before `multi-user.target`, the Debian login prompts will override our own login prompt and info. With `After=multi-user.target` is it assured that Debian login prompts come first (clearing the console and such), then out own login prompt will be assured to show and output is visible on console, including service starts and such.
- However indeed in case of autologin (most `dietpi-autostart` options, but 0) this is indeed suboptimal, since login is done directly after multi-user.target (respectively already with getty@tty1.service), so DietPi-Login for sure will run into the loop waiting for PostBoot to finish.
- I opened this issue to think about a proper solution: https://github.com/Fourdee/DietPi/issues/2476
E.g. Use `After=multi-user.target` only in case of manual console login and remove it if any autologin is active. Also skip our login prompt then, since it is obsolete.

MichaIng commented:
@Fourdee 
Jep was thinking about skipping the forced reboot after update as well. On my systems I always skip reboot (dietpi-update 1) anyway. Only thing that we need to think about: `dietpi-globals` are not reloaded to console session and from within the script we have no change to do this. Logout/login would be sufficient, but AFAIK from within the script we cannot trigger a logout as well?

It is not a very large issue. In current session old G_* functions are available until fresh login. Not sure if it might be an issue state old G_* variables are present as well, e.g. $G_DIETPI_VERSION* and $G_DIETPI_INSTALL_STAGE. But important this is only within scripts which reload the updated globals, thus reread those variables correctly.
